lib/spack/spack/util/parallel.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+from __future__ import print_function
+
+import contextlib
+import multiprocessing
+import os
+import sys
+import traceback
+
+import six
+
+
+class ErrorFromWorker(object):
+    """Wrapper class to report an error from a worker process"""
+    def __init__(self):
+        """Create an error object from an exception raised from
+        the worker process.
+
+        The attributes of the process error objects are all strings
+        as they are easy to send over a pipe.
+
+        Args:
+            exc: exception raised from the worker process
+        """
+        self.pid = os.getpid()
+        exc_cls, exc, tb = sys.exc_info()
+        self.error_message = "{0.__name__}: {1!s}".format(exc_cls, exc)
+        buffer = six.StringIO()
+        traceback.print_tb(tb, file=buffer)
+        self.exc_stacktrace = buffer.getvalue()
+        buffer.close()
+
+    def __str__(self):
+        msg = ("[PID={0.pid}] {0.error_message}\n"
+               "TASK STACKTRACE:\n"
+               "{0.exc_stacktrace}")
+        return msg.format(self)
```

Comment:
I really don't like calling sys.exc_info outside of the except block, it feels like an opportunity to write a bug later.

I also think we can use `traceback.format_exception` to get output much more simply.

```suggestion
    def __init__(self, exc_cls, exc, tb):
        """Create an error object from an exception raised from
        the worker process.

        The attributes of the process error objects are all strings
        as they are easy to send over a pipe.

        Args:
            exc_cls: class of exception raised from worker process
            exc: exception raised from the worker process
            tb: traceback of exception raised from worker process
        """
        self.pid = os.getpid()
        self.exc_cls, self.exc, self.tb = exc_cls, exc, tb

    def __str__(self):
        msg = "Exception from process {0}\n".format(self.pid)
        msg += ''.join(traceback.format_exception(exc_cls, exc, tb))
        return msg
```

Reply:
I tried to keep the spirit of this comment, with the caveat that I'm only storing strings as attributes (might be `multiprocessing` PTSD but if I have to send an object over a pipe, I'm trying to use only built-in types).

Reply:
Done in b809264 Let me know if that's fine with you.